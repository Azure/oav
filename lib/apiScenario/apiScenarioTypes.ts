import { Operation, SwaggerExample } from "../swagger/swaggerTypes";

//#region Common

type TransformRaw<T, Additional = {}, OptionalKey extends keyof T = never> = {
  [P in OptionalKey]?: T[P];
} &
  {
    [P in Exclude<keyof T, OptionalKey | keyof Additional>]-?: Exclude<T[P], undefined>;
  } &
  Additional;

export type VarType = Variable["type"];

export type VarValue = boolean | number | string | VarValue[] | { [key: string]: VarValue };

export type Variable =
  | BoolVariable
  | IntVariable
  | StringVariable
  | SecureStringVariable
  | ArrayVariable
  | ObjectVariable
  | SecureObjectVariable;

export interface StringVariable {
  type: "string";
  value?: string;
}

export interface SecureStringVariable {
  type: "secureString";
  value?: string;
}

export interface BoolVariable {
  type: "bool";
  value?: boolean;
}

export interface IntVariable {
  type: "int";
  value?: number;
}

export interface ObjectVariable {
  type: "object";
  value?: { [key: string]: VarValue };
  patches?: JsonPatchOp[];
}

export interface SecureObjectVariable {
  type: "secureObject";
  value?: { [key: string]: VarValue };
  patches?: JsonPatchOp[];
}

export interface ArrayVariable {
  type: "array";
  value?: VarValue[];
  patches?: JsonPatchOp[];
}

export interface RawVariableScope {
  variables?: {
    [variableName: string]: string | Variable;
  };
}

export interface VariableScope {
  variables: {
    [variableName: string]: Variable;
  };
  requiredVariables: string[];
  secretVariables: string[];
}

export interface OutputVariables {
  [variableName: string]: {
    type?: VarType;
    fromRequest: string;
    fromResponse: string;
  };
}

export interface ReadmeTag {
  name: string;
  filePath: string;
  tag?: string;
}

//#endregion

//#region Step Base

type RawStepBase = RawVariableScope & {
  step?: string;
  description?: string;
  outputVariables?: OutputVariables;
};

type StepBase = VariableScope & {
  isPrepareStep?: boolean;
  isCleanUpStep?: boolean;
};

export type Step = StepRestCall | StepArmTemplate;
export type RawStep = RawStepOperation | RawStepExample | RawStepArmTemplate | RawStepArmScript;

//#endregion

//#region Step RestCall

export type RawStepExample = RawStepBase & {
  exampleFile: string;
  requestUpdate?: JsonPatchOp[];
  responseUpdate?: JsonPatchOp[];
};

export type RawStepOperation = RawStepBase & {
  operationId: string;
  readmeTag?: string;
  parameters?: { [parameterName: string]: VarValue };
  responses?: SwaggerExample["responses"];
};

export type StepRestCallExample = StepBase & {};

export type StepRestCall = StepBase & {
  type: "restCall";
  step: string;
  description?: string;
  operationId: string;
  operation: Operation;
  exampleFile?: string;
  requestParameters: SwaggerExample["parameters"];
  responseExpected: SwaggerExample["responses"];
  outputVariables?: OutputVariables;
};

//#endregion

//#region Step Arm Deployment Script
export type RawStepArmScript = RawStepBase & {
  armDeploymentScript: string;
  arguments?: string;
  environmentVariables?: Array<{
    name: string;
    value: string;
  }>;
};
//#endregion

//#region Step Arm Template

export type ArmTemplateVariableType =
  | "string"
  | "securestring"
  | "int"
  | "bool"
  | "object"
  | "secureObject"
  | "array";

export type RawStepArmTemplate = RawStepBase & {
  armTemplate: string;
};

export type StepArmTemplate = TransformRaw<
  RawStepArmTemplate,
  StepBase & {
    type: "armTemplateDeployment";
    armTemplatePayload: ArmTemplate;
  },
  "description"
>;

export interface ArmResource {
  name: string;
  apiVersion: string;
  type: string;
  location?: string;
  properties?: object;
}

export type ArmDeploymentScriptResource = ArmResource & {
  type: "Microsoft.Resources/deploymentScripts";
  kind: "AzurePowerShell" | "AzureCLI";
  identity?: {
    type: "UserAssigned";
    userAssignedIdentities: {
      [name: string]: {};
    };
  };
  properties: {
    arguments?: string;
    azPowerShellVersion?: string;
    azCliVersion?: string;
    scriptContent: string;
    forceUpdateTag?: string;
    timeout?: string;
    cleanupPreference?: string;
    retentionInterval?: string;
    environmentVariables?: Array<{
      name: string;
      value?: string;
      secureValue?: string;
    }>;
  };
};

export interface ArmTemplate {
  $schema?: string;
  contentVersion?: string;
  parameters?: {
    [name: string]: {
      type: ArmTemplateVariableType;
      defaultValue?: any;
    };
  };
  outputs?: {
    [name: string]: {
      condition?: string;
      type: ArmTemplateVariableType;
    };
  };
  resources?: ArmResource[];
}

//#endregion

//#region JsonPatchOp

export interface JsonPatchOpAdd {
  add: string;
  value: any;
}

export interface JsonPatchOpRemove {
  remove: string;
  oldValue?: any;
}

export interface JsonPatchOpReplace {
  replace: string;
  value: any;
  oldValue?: any;
}

export interface JsonPatchOpCopy {
  copy: string;
  from: string;
}

export interface JsonPatchOpMove {
  move: string;
  from: string;
}

export interface JsonPatchOpTest {
  test: string;
  value: any;
}

export type JsonPatchOp =
  | JsonPatchOpAdd
  | JsonPatchOpRemove
  | JsonPatchOpReplace
  | JsonPatchOpCopy
  | JsonPatchOpMove
  | JsonPatchOpTest;

//#endregion

//#region Scenario

export type RawScenario = RawVariableScope & {
  scenario?: string;
  shareScope?: boolean;
  description?: string;
  steps: RawStep[];
};

export type Scenario = TransformRaw<
  RawScenario,
  {
    steps: Step[];
    _scenarioDef: ScenarioDefinition;
    _resolvedSteps: Step[];
  } & VariableScope
>;

//#endregion

//#region ScenarioDefinitionFile
export type RawScenarioDefinition = RawVariableScope & {
  scope?: "ResourceGroup";
  prepareSteps?: RawStep[];
  scenarios: RawScenario[];
  cleanUpSteps?: RawStep[];
};

export type ScenarioDefinition = TransformRaw<
  RawScenarioDefinition,
  VariableScope & {
    prepareSteps: Step[];
    scenarios: Scenario[];
    cleanUpSteps: Step[];
    _filePath: string;
  }
>;
//#endregion

//#region Runner specific types
export interface RawReport {
  executions: RawExecution[];
  timings: any;
  variables: { [variableName: string]: Variable };
  testScenarioName?: string;
  metadata: any;
}

export interface RawExecution {
  request: RawRequest;
  response: RawResponse;
  annotation?: any;
}
export interface RawRequest {
  url: string;
  method: string;
  headers: { [key: string]: any };
  body: string;
}

export interface RawResponse {
  statusCode: number;
  headers: { [key: string]: any };
  body: string;
}

export interface TestResources {
  ["test-resources"]: Array<{ [key: string]: string }>;
}

//#endregion
