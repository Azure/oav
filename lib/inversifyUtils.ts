import { Container, interfaces } from "inversify";
import { setDefaultOpts } from "./swagger/loader";

export const TYPES = {
  opts: Symbol("InversifyTYPES.opts"),
};

export const inversifyGetInstance = <T, Opt = {}>(
  claz: interfaces.Newable<T>,
  opts: Opt &
    interfaces.ContainerOptions & {
      container?: Container;
    }
) => {
  if (opts.container === undefined) {
    setDefaultOpts(opts, {
      defaultScope: "Singleton",
      autoBindInjectable: true,
    } as any);
    opts.container = new Container(opts);
  }
  opts.container.bind(TYPES.opts).toConstantValue(opts);
  return opts.container.get(claz);
};
