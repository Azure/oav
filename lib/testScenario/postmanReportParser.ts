import { inject, injectable } from "inversify";
import {
  RequestDefinition,
  ResponseDefinition,
  ItemDefinition,
  Request,
  Response,
  DescriptionDefinition,
} from "postman-collection";
import { FileLoader, FileLoaderOption } from "./../swagger/fileLoader";
import { TYPES } from "./../inversifyUtils";
import { RawExecution, RawReport, RawRequest, RawResponse } from "./testResourceTypes";

interface NewmanReport {
  run: Run;
  environment: any;
  collection: any;
}

interface Run {
  executions: NewmanExecution[];
}

interface NewmanExecution {
  item: ItemDefinition;
  request: RequestDefinition;
  response: ResponseDefinition;
}

export interface NewmanReportParserOption extends FileLoaderOption {
  newmanReportFilePath: string;
  reportOutputFilePath?: string;
}

@injectable()
export class NewmanReportParser {
  private report: RawReport;
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  constructor(
    @inject(TYPES.opts) private opts: NewmanReportParserOption,
    private fileLoader: FileLoader
  ) {
    this.report = { variables: {}, executions: [], metadata: {} };
  }

  public async generateRawReport() {
    const content = await this.fileLoader.load(this.opts.newmanReportFilePath);
    const report = JSON.parse(content) as NewmanReport;
    this.report.metadata.testScenarioFilePath = report.collection.info.description.content; // JSON.parse(report.collection.description.content);
    for (const it of report.run.executions) {
      this.report.executions.push(this.generateExampleItem(it));
    }
    this.report.variables = this.parseVariables(report.environment.values);
    if (this.opts.reportOutputFilePath !== undefined) {
      await this.fileLoader.writeFile(
        this.opts.reportOutputFilePath,
        JSON.stringify(this.report, null, 2)
      );
    }
    return this.report;
  }

  private generateExampleItem(it: NewmanExecution): RawExecution {
    const resp = new Response(it.response);
    const req = new Request(it.request);
    const rawReq = this.parseRequest(req);
    const rawResp = this.parseResponse(resp);
    const annotation = JSON.parse((it.item.description as DescriptionDefinition)?.content || "{}");
    return {
      request: rawReq,
      response: rawResp,
      annotation: annotation,
    };
  }

  private parseRequest(req: Request): RawRequest {
    const ret: RawRequest = {
      url: "",
      method: "",
      headers: [],
      body: "",
    };
    ret.url = req.url.getRaw();
    ret.headers = this.parseHeader(req.headers.toJSON());
    ret.method = req.method;
    ret.body = req.body?.toString() || "";
    return ret;
  }

  private parseResponse(resp: Response): RawResponse {
    const ret: RawResponse = {
      headers: [],
      statusCode: resp.code,
      body: "",
    };
    ret.headers = this.parseHeader(resp.headers.toJSON());

    ret.body = resp.stream?.toString() || "";
    return ret;
  }

  private parseHeader(headers: any[]) {
    const ret: any = {};
    for (const it of headers) {
      ret[it.key] = it.value;

      // Currently only mask bearer token header.
      // For further sensitive data, should add mask module here
      if (it.key === "Authorization") {
        ret[it.key] = "<bearer token>";
      }
    }
    return ret;
  }

  private parseVariables(environment: any[]) {
    const ret: any = {};
    for (const it of environment) {
      if (it.type === "string") {
        ret[it.key] = it.value;
      }
    }
    return ret;
  }
}
