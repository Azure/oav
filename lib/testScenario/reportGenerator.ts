import * as path from "path";
import * as uuid from "uuid";
import * as _ from "lodash";
import { injectable, inject } from "inversify";
import { ExampleQualityValidator } from "../exampleQualityValidator/exampleQualityValidator";
import { setDefaultOpts } from "../swagger/loader";
import { Severity } from "../util/severity";
import {
  findNearestReadmeDir,
  getApiVersionFromSwaggerFile,
  getProviderFromFilePath,
} from "../util/utils";
import { SwaggerAnalyzer } from "./swaggerAnalyzer";
import { DataMasker } from "./dataMasker";
import { defaultQualityReportFilePath } from "./defaultNaming";
import { FileLoader } from "./../swagger/fileLoader";
import { TYPES } from "./../inversifyUtils";
import { SwaggerExample } from "./../swagger/swaggerTypes";
import { TestResourceLoader, TestResourceLoaderOption } from "./testResourceLoader";
import { NewmanReportParser, NewmanReportParserOption } from "./postmanReportParser";
import {
  RawReport,
  RawExecution,
  TestDefinitionFile,
  TestStep,
  JsonPatchOp,
  TestStepRestCall,
} from "./testResourceTypes";
import { VariableEnv } from "./variableEnv";
import { getJsonPatchDiff } from "./diffUtils";
import { BlobUploader, BlobUploaderOption } from "./blobUploader";

interface GeneratedExample {
  exampleFilePath: string;
  step: string;
  operationId: string;
  example: SwaggerExample;
}

interface TestScenarioResult {
  testScenario: {
    testScenarioFilePath: string;
    readmeFilePath?: string;
    swaggerFilePaths: string[];
    tag?: string;

    // New added fields
    environment?: string;
    armEnv?: string;
    subscriptionId?: string;
    fileRoot?: string;
    resourceProvider?: string;
    apiVersion?: string;
    startTime?: string;
    endTime?: string;
    runId?: string;
    repository?: string;
    branch?: string;
    commitHash?: string;
  };
  stepResult: { [step: string]: StepResult };
}

interface StepResult {
  exampleFilePath?: string;
  example?: SwaggerExample;
  operationId: string;
  runtimeError?: HttpError;
  responseDiffResult?: { [statusCode: number]: JsonPatchOp[] };
  liveValidationResult?: any;
  stepValidationResult?: any;
  correlationId?: string;
}

interface HttpError {
  statusCode: number;
  rawExecution: RawExecution;
}

export interface ReportGeneratorOption
  extends NewmanReportParserOption,
    TestResourceLoaderOption,
    BlobUploaderOption {
  testDefFilePath: string;
  reportOutputFilePath?: string;
  runId?: string;
}

@injectable()
export class ReportGenerator {
  private exampleQualityValidator: ExampleQualityValidator;
  private swaggerExampleQualityResult: TestScenarioResult;
  private testDefFile: TestDefinitionFile | undefined;
  private operationIds: Set<string>;
  private rawReport: RawReport | undefined;
  private fileRoot: string;
  private recording: Map<string, RawExecution>;
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  constructor(
    @inject(TYPES.opts) private opts: ReportGeneratorOption,
    private postmanReportParser: NewmanReportParser,
    private testResourceLoader: TestResourceLoader,
    private fileLoader: FileLoader,
    private blobUploader: BlobUploader,
    private dataMasker: DataMasker,
    private swaggerAnalyzer: SwaggerAnalyzer
  ) {
    setDefaultOpts(this.opts, {
      newmanReportFilePath: "",
      reportOutputFilePath: defaultQualityReportFilePath(this.opts.newmanReportFilePath),
      enableBlobUploader: false,
      blobConnectionString: "",
      testDefFilePath: "",
      runId: uuid.v4(),
    });
    const swaggerFileAbsolutePaths = this.opts.swaggerFilePaths!.map((it) => path.resolve(it));
    this.exampleQualityValidator = ExampleQualityValidator.create({
      swaggerFilePaths: [...swaggerFileAbsolutePaths],
    });
    this.recording = new Map<string, RawExecution>();
    this.operationIds = new Set<string>();
    this.fileRoot = findNearestReadmeDir(this.opts.testDefFilePath) || "/";

    this.swaggerExampleQualityResult = {
      testScenario: {
        testScenarioFilePath: path.relative(this.fileRoot, this.opts.testDefFilePath),
        swaggerFilePaths: this.opts.swaggerFilePaths!.map((it) => path.relative(this.fileRoot, it)),
        resourceProvider: getProviderFromFilePath(this.opts.testDefFilePath),
        apiVersion: getApiVersionFromSwaggerFile(this.opts.swaggerFilePaths![0]),
        runId: this.opts.runId,
        fileRoot: this.fileRoot,
        repository: process.env.SPEC_REPOSITORY,
        branch: process.env.SPEC_BRANCH,
        commitHash: process.env.COMMIT_HASH,
      },
      stepResult: {},
    };
    this.testDefFile = undefined;
  }

  public async initialize() {
    this.rawReport = await this.postmanReportParser.generateRawReport();
  }

  public async generateExampleQualityReport(rawReport: RawReport) {
    await this.initialize();
    const variables = rawReport.variables;
    this.swaggerExampleQualityResult.testScenario.startTime = rawReport.timings.started;
    this.swaggerExampleQualityResult.testScenario.endTime = rawReport.timings.completed;
    this.swaggerExampleQualityResult.testScenario.subscriptionId = variables.subscriptionId;
    for (const it of rawReport.executions) {
      if (it.annotation === undefined) {
        continue;
      }
      if (it.annotation.type === "simple" || it.annotation.type === "LRO") {
        let error: any = {};
        const generatedExample = this.generateExample(it, variables, rawReport);
        const matchedStep = this.getMatchedStep(it.annotation.step) as TestStepRestCall;
        if (
          Math.floor(it.response.statusCode / 200) !== 1 &&
          it.response.statusCode !== matchedStep.statusCode
        ) {
          error = { statusCode: it.response.statusCode, rawExecution: it } as HttpError;
        }
        if (matchedStep === undefined) {
          continue;
        }
        // validate real payload.
        const roundtripErrors = await this.exampleQualityValidator.validateExternalExamples([
          {
            exampleFilePath: generatedExample.exampleFilePath,
            example: generatedExample.example,
            operationId: matchedStep.operationId,
          },
        ]);
        const correlationId = it.response.headers["x-ms-correlation-request-id"];
        this.swaggerExampleQualityResult.stepResult[it.annotation.step] = {
          exampleFilePath: generatedExample.exampleFilePath,
          operationId: it.annotation.operationId,
          runtimeError: error,
          responseDiffResult: this.exampleResponseDiff(generatedExample, matchedStep),
          stepValidationResult: roundtripErrors,
          correlationId: correlationId,
        };
        this.recording.set(correlationId, it);
      }
    }
    if (this.opts.reportOutputFilePath !== undefined) {
      console.log(`Write generated report file: ${this.opts.reportOutputFilePath}`);
      await this.fileLoader.writeFile(
        this.opts.reportOutputFilePath,
        JSON.stringify(this.swaggerExampleQualityResult, null, 2)
      );
      if (this.opts.enableBlobUploader) {
        const provider = getProviderFromFilePath(this.opts.reportOutputFilePath) || "";
        const idx = this.opts.reportOutputFilePath.indexOf(provider);
        const blobPath = this.opts.reportOutputFilePath.substr(
          idx,
          this.opts.blobConnectionString?.length
        );
        const secretValues: string[] = [];
        for (const [k, v] of Object.entries(this.rawReport?.variables)) {
          if (this.dataMasker.maybeSecretKey(k)) {
            secretValues.push(v as string);
          }
        }
        this.dataMasker.addMaskedValues(secretValues);
        this.dataMasker.addMaskedKeys(await this.swaggerAnalyzer.getAllSecretKey());

        // mask report here.
        await this.blobUploader.uploadContent(
          "report",
          blobPath,
          this.dataMasker.jsonStringify(this.swaggerExampleQualityResult)
        );

        for (const [correlationId, v] of this.recording) {
          const payloadBlobPath = `${path.dirname(blobPath)}/${correlationId}.json`;
          await this.blobUploader.uploadContent(
            "payload",
            payloadBlobPath,
            JSON.stringify(v, null, 2)
          );
        }
      }
    }
    console.log(JSON.stringify(this.swaggerExampleQualityResult, null, 2));
  }

  private generateExample(
    it: RawExecution,
    variables: any,
    rawReport: RawReport
  ): GeneratedExample {
    const example: any = {};
    if (it.annotation.operationId !== undefined) {
      this.operationIds.add(it.annotation.operationId);
    }
    example.parameters = this.generateParametersFromQuery(variables, it);
    try {
      _.extend(example.parameters, { parameters: JSON.parse(it.request.body) });
      // eslint-disable-next-line no-empty
    } catch (err) {}
    const resp: any = this.parseRespBody(it);
    example.responses = {};
    _.extend(example.responses, resp);
    const exampleFilePath = path.relative(
      this.fileRoot,
      path.resolve(this.opts.testDefFilePath, it.annotation.exampleName)
    );
    const generatedGetExecution = this.findGeneratedGetExecution(it, rawReport);
    if (generatedGetExecution.length > 0) {
      const getResp = this.parseRespBody(generatedGetExecution[0]);
      _.extend(example.responses, getResp);
    }
    return {
      exampleFilePath: exampleFilePath,
      example,
      step: it.annotation.step,
      operationId: it.annotation.operationId,
    };
  }

  private exampleResponseDiff(
    example: GeneratedExample,
    matchedStep: TestStep
  ): { [statusCode: number]: JsonPatchOp[] } {
    const res: any = {};
    if (matchedStep?.type === "restCall") {
      if (example.example.responses[matchedStep.statusCode] !== undefined) {
        res[matchedStep.statusCode] = this.responseDiff(
          example.example.responses[matchedStep.statusCode]?.body || {},
          matchedStep.responseExpected,
          this.rawReport!.variables
        );
      }
    }
    return res;
  }

  private responseDiff(resp: any, expectedResp: any, variables: any) {
    const env = new VariableEnv();
    env.setBatch(variables);
    try {
      const expected = env.resolveObjectValues(expectedResp);
      const delta = getJsonPatchDiff(expected, resp, {
        includeOldValue: true,
        minimizeDiff: false,
      }).map((it: any) => {
        const ret: any = {};
        if (it.remove !== undefined) {
          ret.code = "RESPONSE_ADDITIONAL_VALUE";
          ret.severity = Severity.Error;
        } else if (it.add !== undefined) {
          ret.code = "RESPONSE_MISSING_VALUE";
          ret.severity = Severity.Error;
        } else {
          ret.code = "RESPONSE_INCORRECT_VALUE";
          ret.severity = Severity.Error;
        }
        ret.detail = it;
        return ret;
      });
      return delta;
    } catch (err) {
      console.log(err);
    }
    return [];
  }

  private getMatchedStep(stepName: string): TestStep | undefined {
    for (const it of this.testDefFile?.prepareSteps ?? []) {
      if (stepName === it.step) {
        return it;
      }
    }
    for (const testScenario of this.testDefFile?.testScenarios ?? []) {
      for (const step of testScenario.steps) {
        if (stepName === step.step) {
          return step;
        }
      }
    }
    return undefined;
  }
  private findGeneratedGetExecution(it: RawExecution, rawReport: RawReport) {
    if (it.annotation.type === "LRO") {
      const finalGet = rawReport.executions.filter(
        (execution) =>
          execution.annotation &&
          execution.annotation.lro_item_name === it.annotation.itemName &&
          execution.annotation.type === "generated-get"
      );
      return finalGet;
    }
    return [];
  }

  public async generateReport() {
    this.testDefFile = await this.testResourceLoader.load(this.opts.testDefFilePath);
    await this.initialize();
    await this.generateExampleQualityReport(this.rawReport!);
  }

  private parseRespBody(it: RawExecution) {
    const resp: any = {};
    try {
      resp[it.response.statusCode] = { body: JSON.parse(it.response.body) };
    } catch (err) {
      resp[it.response.statusCode] = { body: it.response.body };
    }
    return resp;
  }

  private generateParametersFromQuery(variables: any, execution: RawExecution) {
    const ret: any = {};
    for (const [k, v] of Object.entries(variables)) {
      if (typeof v === "string") {
        if (execution.request.url.includes(v as string)) {
          ret[k] = v;
        }
      }
    }
    return ret;
  }
}
