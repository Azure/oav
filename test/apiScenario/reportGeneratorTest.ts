import { inversifyGetInstance } from "../../lib/inversifyUtils";
import { NewmanReportParser } from "../../lib/apiScenario/postmanReportParser";
import { RawReport } from "../../lib/apiScenario/apiScenarioTypes";
import { ReportGenerator, ReportGeneratorOption } from "../../lib/apiScenario/reportGenerator";

describe("reportGenerator", () => {
  it("generate report - storage", async () => {
    const newmanReportFilePath =
      "test/apiScenario/fixtures/report/storageBasicExample.yaml/202207041617-tywsob/StorageBasicExample.json";
    const newmanReportParser = inversifyGetInstance(NewmanReportParser, {
      newmanReportFilePath,
    });
    const rawReport: RawReport = await newmanReportParser.generateRawReport(newmanReportFilePath);
    const apiScenarioFilePath = rawReport.metadata.apiScenarioFilePath;
    const reportGeneratorOption: ReportGeneratorOption = {
      newmanReportFilePath,
      apiScenarioFilePath,
      apiScenarioName: rawReport.metadata.apiScenarioName,
      swaggerFilePaths: rawReport.metadata.swaggerFilePaths,
      checkUnderFileRoot: false,
      eraseXmsExamples: false,
      eraseDescription: false,
      runId: "",
    };
    const reportGenerator = inversifyGetInstance(ReportGenerator, reportGeneratorOption);
    reportGeneratorOption.reportOutputFilePath = undefined;
    const report = await reportGenerator.generateReport();
    report.rootPath = "";
    report.stepResult.forEach((stepResult) => {
      stepResult.liveValidationResult?.requestValidationResult.errors.forEach((error) => {
        error.source = {
          ...error.source,
          url: "",
        };
      });
      stepResult.liveValidationResult?.responseValidationResult.errors.forEach((error) => {
        error.source = {
          ...error.source,
          url: "",
        };
      });
    });

    expect(report).toMatchSnapshot();
  });
});
