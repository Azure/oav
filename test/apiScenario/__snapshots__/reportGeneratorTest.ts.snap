// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reportGenerator generate report - storage 1`] = `
Object {
  "apiVersion": "2021-08-01",
  "armEndpoint": "https://management.azure.com",
  "branch": undefined,
  "commitHash": undefined,
  "endTime": "2022-07-04T08:17:19.575Z",
  "environment": "test",
  "providerNamespace": "Microsoft.Storage",
  "repository": undefined,
  "rootPath": "",
  "runId": "",
  "startTime": "2022-07-04T08:17:09.368Z",
  "stepResult": Array [
    Object {
      "correlationId": "d6658422-9eff-4e4f-ac2a-b154df45d26a",
      "exampleFilePath": "../Microsoft.Storage/stable/2021-08-01/scenarios/examples/StorageAccountCheckNameAvailability.json",
      "liveValidationResult": Object {
        "requestValidationResult": Object {
          "errors": Array [],
          "isSuccessful": true,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_CheckNameAvailability",
          },
          "runtimeException": undefined,
        },
        "responseValidationResult": Object {
          "errors": Array [],
          "isSuccessful": true,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_CheckNameAvailability",
          },
          "runtimeException": undefined,
        },
      },
      "operationId": "StorageAccounts_CheckNameAvailability",
      "responseDiffResult": Array [],
      "runtimeError": Array [],
      "statusCode": 200,
      "stepName": "checkName",
      "stepValidationResult": Array [],
    },
    Object {
      "correlationId": "b69b9ae0-c247-485d-97ed-24940b40191b",
      "exampleFilePath": "../Microsoft.Storage/stable/2021-08-01/scenarios/examples/StorageAccountCreate.json",
      "liveValidationResult": Object {
        "requestValidationResult": Object {
          "errors": Array [],
          "isSuccessful": true,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_Create",
          },
          "runtimeException": undefined,
        },
        "responseValidationResult": Object {
          "errors": Array [
            Object {
              "code": "INVALID_RESPONSE_CODE",
              "documentationUrl": "",
              "jsonPathsInPayload": Array [],
              "message": "The swagger file does not define '400' response code",
              "pathsInPayload": Array [],
              "schemaPath": "",
              "severity": 0,
              "source": Object {
                "position": Object {
                  "column": 22,
                  "line": 180,
                },
                "url": "",
              },
            },
          ],
          "isSuccessful": false,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_Create",
          },
          "runtimeException": undefined,
        },
      },
      "operationId": "StorageAccounts_Create",
      "responseDiffResult": Array [],
      "runtimeError": Array [
        Object {
          "code": "RUNTIME_ERROR",
          "detail": "{\\"error\\":{\\"code\\":\\"UnsupportedEdgeZone\\",\\"message\\":\\"Edge zone 'losangeles001' not found. The available edge zones in location 'westus' are ''.\\"}}",
          "message": "statusCode: 400, code: UnsupportedEdgeZone, message: Edge zone 'losangeles001' not found. The available edge zones in location 'westus' are ''.",
          "severity": "Error",
        },
      ],
      "statusCode": 400,
      "stepName": "createStorageAccount",
      "stepValidationResult": Array [],
    },
    Object {
      "correlationId": "a1187deb-0f76-4958-a75a-4e74c6895c8a",
      "exampleFilePath": "../Microsoft.Storage/stable/2021-08-01/scenarios/examples/StorageAccountUpdate.json",
      "liveValidationResult": Object {
        "requestValidationResult": Object {
          "errors": Array [],
          "isSuccessful": true,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_Update",
          },
          "runtimeException": undefined,
        },
        "responseValidationResult": Object {
          "errors": Array [
            Object {
              "code": "INVALID_RESPONSE_CODE",
              "documentationUrl": "",
              "jsonPathsInPayload": Array [],
              "message": "The swagger file does not define '404' response code",
              "pathsInPayload": Array [],
              "schemaPath": "",
              "severity": 0,
              "source": Object {
                "position": Object {
                  "column": 22,
                  "line": 334,
                },
                "url": "",
              },
            },
          ],
          "isSuccessful": false,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_Update",
          },
          "runtimeException": undefined,
        },
      },
      "operationId": "StorageAccounts_Update",
      "responseDiffResult": Array [],
      "runtimeError": Array [
        Object {
          "code": "RUNTIME_ERROR",
          "detail": "{\\"error\\":{\\"code\\":\\"PatchResourceNotFound\\",\\"target\\":\\"staj8625u\\",\\"message\\":\\"The resource '/subscriptions/db5eb68e-73e2-4fa8-b18a-46cd1be4cce5/resourceGroups/apiTest-ep9iua/providers/Microsoft.Storage/storageAccounts/staj8625u' was not found when evaluating policies for a PATCH operation.\\"}}",
          "message": "statusCode: 404, code: PatchResourceNotFound, message: The resource '/subscriptions/db5eb68e-73e2-4fa8-b18a-46cd1be4cce5/resourceGroups/apiTest-ep9iua/providers/Microsoft.Storage/storageAccounts/staj8625u' was not found when evaluating policies for a PATCH operation.",
          "severity": "Error",
        },
      ],
      "statusCode": 404,
      "stepName": "updateStorageAccount",
      "stepValidationResult": Array [],
    },
    Object {
      "correlationId": "71c47075-2193-48c4-8f77-8da2f66de14a",
      "exampleFilePath": "../Microsoft.Storage/stable/2021-08-01/scenarios/examples/StorageAccountGetProperties.json",
      "liveValidationResult": Object {
        "requestValidationResult": Object {
          "errors": Array [],
          "isSuccessful": true,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_GetProperties",
          },
          "runtimeException": undefined,
        },
        "responseValidationResult": Object {
          "errors": Array [
            Object {
              "code": "INVALID_RESPONSE_CODE",
              "documentationUrl": "",
              "jsonPathsInPayload": Array [],
              "message": "The swagger file does not define '404' response code",
              "pathsInPayload": Array [],
              "schemaPath": "",
              "severity": 0,
              "source": Object {
                "position": Object {
                  "column": 22,
                  "line": 270,
                },
                "url": "",
              },
            },
          ],
          "isSuccessful": false,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_GetProperties",
          },
          "runtimeException": undefined,
        },
      },
      "operationId": "StorageAccounts_GetProperties",
      "responseDiffResult": Array [],
      "runtimeError": Array [
        Object {
          "code": "RUNTIME_ERROR",
          "detail": "{\\"error\\":{\\"code\\":\\"ResourceNotFound\\",\\"message\\":\\"The Resource 'Microsoft.Storage/storageAccounts/staj8625u' under resource group 'apiTest-ep9iua' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix\\"}}",
          "message": "statusCode: 404, code: ResourceNotFound, message: The Resource 'Microsoft.Storage/storageAccounts/staj8625u' under resource group 'apiTest-ep9iua' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix",
          "severity": "Error",
        },
      ],
      "statusCode": 404,
      "stepName": "getStorageAccount",
      "stepValidationResult": Array [],
    },
    Object {
      "correlationId": "fd1c31fb-ef89-4671-8865-6b8aeb7cd1c9",
      "exampleFilePath": "../Microsoft.Storage/stable/2021-08-01/scenarios/examples/StorageAccountDelete.json",
      "liveValidationResult": Object {
        "requestValidationResult": Object {
          "errors": Array [],
          "isSuccessful": true,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_Delete",
          },
          "runtimeException": undefined,
        },
        "responseValidationResult": Object {
          "errors": Array [],
          "isSuccessful": true,
          "operationInfo": Object {
            "apiVersion": "2021-08-01",
            "operationId": "StorageAccounts_Delete",
          },
          "runtimeException": undefined,
        },
      },
      "operationId": "StorageAccounts_Delete",
      "responseDiffResult": Array [],
      "runtimeError": Array [],
      "statusCode": 204,
      "stepName": "deleteStorageAccount",
      "stepValidationResult": Array [],
    },
  ],
  "subscriptionId": "aaaaa",
  "swaggerFilePaths": Array [],
  "testScenarioFilePath": "../Microsoft.Storage/stable/2021-08-01/scenarios/storageBasicExample.yaml",
  "testScenarioName": "StorageBasicExample",
}
`;
